aiohttp_s3_client/credentials.py

```diff
@@ -0,0 +1,317 @@
+            result[field] = value
+
+    return StaticCredentials(**result)
+
+
+def collect_credentials(
+    *, url: Optional[URL] = None, **kwargs,
+) -> StaticCredentials:
+    credentials: List[StaticCredentials] = []
+    if kwargs:
+        credentials.append(StaticCredentials(**kwargs))
+    if url:
+        credentials.append(URLCredentials(url))
+    credentials.append(EnvironmentCredentials())
+    credentials.append(ConfigCredentials())
+    return merge_credentials(*credentials)
+
+
+class MetadataDocument(TypedDict, total=False):
+    """
+    Response example is:
+
+        {
+          "accountId" : "123123",
+          "architecture" : "x86_64",
+          "availabilityZone" : "us-east-1a",
+          "billingProducts" : null,
+          "devpayProductCodes" : null,
+          "marketplaceProductCodes" : null,
+          "imageId" : "ami-123123",
+          "instanceId" : "i-11232323",
+          "instanceType" : "t3a.micro",
+          "kernelId" : null,
+          "pendingTime" : "2023-06-13T18:18:58Z",
+          "privateIp" : "172.33.33.33",
+          "ramdiskId" : null,
+          "region" : "us-east-1",
+          "version" : "2017-09-30"
+        }
+    """
+    region: str
+
+
+class MetadataSecurityCredentials(TypedDict, total=False):
+    Code: str
+    Type: str
+    AccessKeyId: str
+    SecretAccessKey: str
+    Token: str
+    Expiration: str
+
+
+class MetadataCredentials(AbstractCredentials):
+    METADATA_ADDRESS = "169.254.169.254"
+    METADATA_PORT = 80
+
+    def __bool__(self) -> bool:
+        """ Fast check metadata server connection """
+
+        with socket.socket(
+            socket.AF_INET,
+            socket.SOCK_STREAM,
+        ) as sock:
+            return sock.connect_ex(
+                (self.METADATA_ADDRESS, self.METADATA_PORT),
+            ) == 0
+
+    def __init__(self, *, service: str = "s3"):
+        self.session = aiohttp.ClientSession(
+            base_url=URL.build(
+                scheme="http",
+                host=self.METADATA_ADDRESS,
+                port=self.METADATA_PORT,
+            ),
+            headers={},
+        )
+        self.service = service
+        self.refresh_lock: asyncio.Lock = asyncio.Lock()
+        self._signer: Optional[AwsRequestSigner] = None
+        self._tasks: List[asyncio.Task] = []
+
+    async def _refresher(self) -> None:
+        while True:
+            async with self.refresh_lock:
+                try:
+                    credentials, expires_at = await self._fetch_credentials()
+                    self._signer = AwsRequestSigner(**credentials.as_dict())
+                    delta = expires_at - datetime.datetime.utcnow()
+                    sleep_time = math.floor(delta.total_seconds() / 2)
+                except Exception:
+                    log.exception("Failed to update credentials")
+                    sleep_time = 60
+            await asyncio.sleep(sleep_time)
+
+    async def start(self) -> None:
```

Comment:
maybe a 
`self.is_started = loop.create_future()`   with a `await self.is_started` after the create_task

Reply:
to clarify, as is it's pretty easy to hit a race condition if you get going right after start()

`RuntimeError: MetadataCredentials must be started before using`